lib/SILGen/ArgumentSource.cpp

```diff
@@ -217,3 +217,17 @@ void ArgumentSource::forwardInto(SILGenFunction &SGF,
   auto substLoweredType = destTL.getLoweredType().getSwiftRValueType();
   RValue(SGF, loc, substLoweredType, outputValue).forwardInto(SGF, loc, dest);
 }
+
+SILType ArgumentSource::getSILSubstRValueType(SILGenFunction &SGF) const & {
+  CanSILFunctionType funcType = SGF.F.getLoweredFunctionType();
+  CanType substType = getSubstType();
+  AbstractionPattern origType(funcType->getGenericSignature(), substType);
```

Comment:
You could add an assert that when the argument source is eventually evaluated down to a single managed value, the type of that matches the expected SIL type.